### PR TITLE
Add basic model-parallel interfaces

### PR DESCRIPTION
**Pull request by levelfour**
## Overview

This P/R adds the most basic interfaces for model parallelization:

+ `chainermn.functions.Send`
+ `chainermn.functions.Recv`

to enable forward/backprop between multiple machines.

## Potential Problem

Above functions are based on `chainermn.communicators.CommunicatorBase.{send,recv}`. They use mpi4py fast communication (cf. `NumPy arrays` in [Tutorial - mpi4py](http://mpi4py.readthedocs.io/en/stable/tutorial.html#point-to-point-communication)), which discards any information attached to `chainer.Variable` objects. This might cause some potential problems, but I am still not aware of them. Please be sure when you use these experimental features.

### CODE REVIEW

File: chainermn/communicators/_base.py [hunk 1]
```diff
@@ -15,6 +19,19 @@ def rank(self):
     def size(self):
         return self.mpi_comm.size
 
+    def send(self, array, dest, tag):
+        shape = numpy.array(array.shape, dtype='int')
```
Comment (iwiwi) [inline chainermn/communicators/_base.py]: If there is no good reason, how about using `np.int32` instead of using type strings? We are consistently using dtype objects (e.g., `np.int32`) in ChainerMN. FYI: `cupy.int32 == numpy.int32`

File: chainermn/communicators/_base.py [hunk 2]
```diff
@@ -15,6 +19,19 @@ def rank(self):
...
+        buf = _memory_utility.array_to_buffer_object(array.astype('float32'))
```
Comment (iwiwi) [inline chainermn/communicators/_base.py]: As forcing types is a little dangerous, how about check that the input array is `float32` by `assert`?

File: chainermn/communicators/_base.py [hunk 3]
```diff
@@ -15,6 +19,19 @@ def rank(self):
...
+        self.mpi_comm.Send([shape, mpi4py.MPI.INT], dest=dest, tag=tag)
+        self.mpi_comm.Send(buf, dest=dest, tag=tag)
+
+    def recv(self, source, tag):
+        shape = numpy.empty(2, dtype='int')
+        self.mpi_comm.Recv([shape, mpi4py.MPI.INT], source=source, tag=tag)
+        buf = numpy.empty(shape[0] * shape[1], dtype='float32')
```
Comment (iwiwi) [inline chainermn/communicators/_base.py]: Seemingly this code only supports two dimensional arrays. We need to handle arrays of different dimensions. For example, when using CNNs, activation is four dimensional arrays.

File: chainermn/functions/communication.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import chainer
+from chainer import cuda
+
+
+class Send(chainer.Function):
+    def __init__(self, comm, peer_rank, peer_tag, device=-1):
+        self.comm = comm
+        self.peer_rank = peer_rank
+        self.peer_tag = peer_tag
+        self.device = device
```
Comment (iwiwi) [inline chainermn/functions/communication.py]: never used?

File: chainermn/functions/__init__.py [hunk 5]
```diff
@@ -0,0 +1,2 @@
+from chainermn.functions.communication import Recv  # NOQA
+from chainermn.functions.communication import Send  # NOQA
```
Comment (iwiwi) [inline chainermn/functions/__init__.py]: Module name `communication` seems too general to me, as ChainerMN is almost all about communication 😉. How about thinking more specific names, such as `point_to_point_communication` ?

File: tests/test_communicator.py [hunk 6]
```diff
@@ -88,6 +88,18 @@ def test_size(self):
         self.assertEqual(self.communicator.size,
                          self.mpi_comm.Get_size())
 
+    def test_send_and_recv(self):
+        if self.communicator.size < 2:
+            raise nose.plugins.skip.SkipTest()
+
+        rank_next = (self.communicator.rank + 1) % self.communicator.size
+        rank_prev = (self.communicator.rank - 1) % self.communicator.size
+        data_send = self.communicator.rank * np.ones((10, 5))
```
Comment (iwiwi) [inline tests/test_communicator.py]: Please try arrays of different dimensions

**Comment by iwiwi** [comment]
How about labeling these functions by `chaienr.utils.experimental` first? c.f., https://github.com/pfnet/chainer/blob/83cff16e237f4e1d2b1e7762ad43a7f7b9e47e7e/chainer/links/theano/theano_function.py#L65

File: chainermn/communicators/_base.py (hunk 3 above)
Comment (levelfour) [inline chainermn/communicators/_base.py]: To handle this, let them first send ndim before shape.

File: chainermn/functions/point_to_point_communication.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import chainer
+from chainer import cuda
+import chainer.utils
+
+
+class Send(chainer.Function):
```
Comment (iwiwi) [inline chainermn/functions/point_to_point_communication.py]: Could you write docstring? Please refer to those of `Function` in Chainer

File: .travis.yml [hunk 8]
```diff
@@ -52,7 +52,7 @@ script:
   - flake8 --config=.flake8.cython .
   - autopep8 -r . --global-config .pep8 | tee check_autopep8
   - test ! -s check_autopep8
-  - for NP in 1 2; do PYTHONWARNINGS='ignore::FutureWarning,module::DeprecationWarning' mpiexec -n ${NP} nosetests -a '!nccl,!gpu'; done
+  - for NP in 1 2; do PYTHONWARNINGS='ignore::FutureWarning,module::DeprecationWarning' mpiexec -n ${NP} nosetests -v -a '!nccl,!gpu'; done
```
Comment (iwiwi) [inline .travis.yml]: Do we need to add `-v` ?

File: chainermn/communicators/_base.py [hunk 9]
```diff
@@ -15,6 +20,34 @@ def rank(self):
     def size(self):
         return self.mpi_comm.size
 
+    def send(self, array, dest, tag):
```
Comment (iwiwi) [inline chainermn/communicators/_base.py]: Could you write docstrings?

File: .travis.yml (hunk 8 above)
Comment (levelfour) [inline .travis.yml]: I think it would be better to add because we can check which testcase fails in case of failure on Travis.
If there's some trouble in combination with Travis, I'll remove it.

File: chainermn/functions/point_to_point_communication.py (hunk 7 above)
Comment (levelfour) [inline chainermn/functions/point_to_point_communication.py]: I added in the latest commit.

File: chainermn/communicators/_base.py (hunk 9 above)
Comment (levelfour) [inline chainermn/communicators/_base.py]: added.

File: chainermn/functions/__init__.py [hunk 10]
```diff
@@ -0,0 +1,4 @@
+from chainermn.functions.point_to_point_communication import recv  # NOQA
+from chainermn.functions.point_to_point_communication import Recv  # NOQA
+from chainermn.functions.point_to_point_communication import send  # NOQA
+from chainermn.functions.point_to_point_communication import Send  # NOQA
```
Comment (iwiwi) [inline chainermn/functions/__init__.py]: Do we need to expose `Recv` and `Send` (in addition to `recv` and `send`)? For example, in Chainer, `BatchNormalizationFunction` is not exposed, and only method `batch_normalization` is imported in `__init__.py`.
https://github.com/chainer/chainer/blob/master/chainer/functions/__init__.py#L256
https://github.com/chainer/chainer/blob/master/chainer/functions/normalization/batch_normalization.py

File: chainermn/communicators/_base.py [hunk 11]
```diff
@@ -15,6 +20,57 @@ def rank(self):
...
+        """A primitive feature of inter-process transmission.
+
+        This method tries to receive numpy-array from target process.
+        This method relies on mpi4py fast communication optimized for
+        numpy arrays, which discards any information attached to
+        chainer.Variable objects. Please be sure.
+
+        Args:
+            source (int): Target process specifier.
+            tag (int): An optional message ID.
```
Comment (iwiwi) [inline chainermn/communicators/_base.py]: What do you mean by 'optional'? It looks a mandatory argument.

File: chainermn/communicators/_base.py [hunk 12]
```diff
@@ -15,6 +20,57 @@ def rank(self):
...
+        buf = _memory_utility.array_to_buffer_object(array)
+        self.mpi_comm.Send([ndim, mpi4py.MPI.INT], dest=dest, tag=tag)
+        self.mpi_comm.Send([shape, mpi4py.MPI.INT], dest=dest, tag=tag)
+
+        if chainer.cuda.get_array_module(array) is not numpy:
+            chainer.cuda.Stream.null.synchronize()
+        self.mpi_comm.Send(buf, dest=dest, tag=tag)
+
+    def recv(self, source, tag):
+        """A primitive feature of inter-process transmission.
```
Comment (iwiwi) [inline chainermn/communicators/_base.py]: The summary lines for `send` and `recv` is the same, could you make it a little more descriptive?

File: chainermn/communicators/_base.py (hunk 11 above)
Comment (levelfour) [inline chainermn/communicators/_base.py]: Actually this comment seems confusing. I updated.

File: chainermn/functions/__init__.py (hunk 10 above)
Comment (levelfour) [inline chainermn/functions/__init__.py]: At the moment I do not feel they should be exposed, so I hid them in the latest commit.

**Comment by levelfour** [comment]
At the moment I added `chainermn.functions.send` and `chainermn.functions.recv` to the API reference.
